File: examples/playwright/python/per_context_fingerprint.py
"""
PRIVACY RESEARCH USE ONLY
Run exclusively in authorized privacy research labs that comply with all applicable laws.
See: https://github.com/botswin/BotBrowser/blob/main/DISCLAIMER.md

BotBrowser Per-Context Fingerprint + Proxy Example (Python Playwright)

Creates multiple browser contexts in a single browser instance, each with
a different fingerprint profile and optional proxy via BotBrowser.setBrowserContextFlags.

Requirements:
    pip install playwright
    BotBrowser binary (not stock Chromium)
    At least two .enc profile files

Usage:
    BOTBROWSER_EXEC_PATH=/path/to/chrome \
    PROFILE_A=/path/to/profile-a.enc \
    PROFILE_B=/path/to/profile-b.enc \
    PROXY_B=socks5://user:pass@host:1080 \
    python per_context_fingerprint.py
"""

import asyncio
import json
import os
import sys

from playwright.async_api import async_playwright


async def create_bb_context(browser, cdp_session, profile_path, proxy=None):
    """
    Create a BrowserContext with a BotBrowser fingerprint profile and optional proxy.

    Steps:
        1. Snapshot existing context IDs
        2. Create context via Playwright
        3. Find the new context ID by diffing before/after
        4. Set BotBrowser flags BEFORE creating any page
    """
    # Step 1: snapshot
    result = await cdp_session.send("Target.getBrowserContexts")
    before_ids = result["browserContextIds"]

    # Step 2: create context
    context = await browser.new_context()

    # Step 3: find new context ID
    result = await cdp_session.send("Target.getBrowserContexts")
    after_ids = result["browserContextIds"]
    context_id = None
    for cid in after_ids:
        if cid not in before_ids:
            context_id = cid
            break

    if not context_id:
        raise RuntimeError("Could not determine browserContextId")

    # Step 4: set flags (profile + optional proxy)
    flags = [f"--bot-profile={profile_path}"]
    if proxy:
        flags.append(f"--proxy-server={proxy}")

    await cdp_session.send("BotBrowser.setBrowserContextFlags", {
        "browserContextId": context_id,
        "botbrowserFlags": flags,
    })

    return context, context_id


async def collect_fingerprint(page):
    """Collect key fingerprint signals from a page."""
    return await page.evaluate("""async () => {
        const r = {};
        r.userAgent = navigator.userAgent;
        r.platform = navigator.platform;
        r.hardwareConcurrency = navigator.hardwareConcurrency;
        r.deviceMemory = navigator.deviceMemory;
        r.screenWidth = screen.width;
        r.screenHeight = screen.height;
        r.timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
        r.languages = JSON.stringify(navigator.languages);
        try {
            const c = document.createElement('canvas');
            const gl = c.getContext('webgl');
            if (gl) {
                const d = gl.getExtension('WEBGL_debug_renderer_info');
                if (d) r.webglRenderer = gl.getParameter(d.UNMASKED_RENDERER_WEBGL);
            }
        } catch (_) {}
        return r;
    }""")


def print_fingerprint(label, fp):
    print(f"\n  [{label}]")
    print(f"    UA:       {fp['userAgent'][:80]}")
    print(f"    Platform: {fp['platform']}")
    print(f"    HW:       {fp['hardwareConcurrency']} cores, {fp.get('deviceMemory', 'N/A')}GB")
    print(f"    Screen:   {fp['screenWidth']}x{fp['screenHeight']}")
    print(f"    WebGL:    {fp.get('webglRenderer', 'N/A')}")
    print(f"    Timezone: {fp['timezone']}")
    print(f"    Langs:    {fp.get('languages', 'N/A')}")


async def main():
    exec_path = os.environ.get("BOTBROWSER_EXEC_PATH")
    profile_a = os.environ.get("PROFILE_A")
    profile_b = os.environ.get("PROFILE_B")
    proxy_b = os.environ.get("PROXY_B")  # optional

    if not exec_path or not profile_a or not profile_b:
        print("Usage:")
        print("  BOTBROWSER_EXEC_PATH=/path/to/chrome \\")
        print("  PROFILE_A=/path/to/profile-a.enc \\")
        print("  PROFILE_B=/path/to/profile-b.enc \\")
        print("  PROXY_B=socks5://user:pass@host:1080 \\")
        print("  python per_context_fingerprint.py")
        sys.exit(1)

    async with async_playwright() as pw:
        # Launch with profile A as the base (default) profile
        browser = await pw.chromium.launch(
            executable_path=exec_path,
            headless=True,
            args=[
                "--disable-blink-features=AutomationControlled",
                "--disable-audio-output",
                f"--bot-profile={profile_a}",
            ],
        )
        print("Browser launched")

        cdp = await browser.new_browser_cdp_session()

        try:
            # --- Context A: profile A (no proxy) ---
            ctx_a, ctx_id_a = await create_bb_context(browser, cdp, profile_a)
            page_a = await ctx_a.new_page()
            print(f"\n  Context A created (id={ctx_id_a})")

            # --- Context B: profile B + optional proxy ---
            ctx_b, ctx_id_b = await create_bb_context(browser, cdp, profile_b, proxy=proxy_b)
            page_b = await ctx_b.new_page()
            print(f"  Context B created (id={ctx_id_b})")
            if proxy_b:
                print(f"  Context B proxy: {proxy_b}")

            # Navigate both
            await asyncio.gather(
                page_a.goto("https://example.com", wait_until="domcontentloaded"),
                page_b.goto("https://example.com", wait_until="domcontentloaded"),
            )

            # Collect fingerprints
            fp_a = await collect_fingerprint(page_a)
            fp_b = await collect_fingerprint(page_b)

            print_fingerprint("Context A", fp_a)
            print_fingerprint("Context B", fp_b)

            # Check proxy IP for context B
            if proxy_b:
                print("\n  --- Proxy check (Context B) ---")
                try:
                    await page_b.goto("https://httpbin.org/ip", timeout=20000)
                    body = await page_b.evaluate("() => document.body.innerText")
                    ip = json.loads(body).get("origin", "N/A")
                    print(f"    Exit IP: {ip}")
                except Exception as e:
                    print(f"    Proxy check failed: {e}")

            # Verify isolation
            print("\n  --- Isolation check ---")
            fields = ["userAgent", "platform", "hardwareConcurrency", "screenWidth", "webglRenderer"]
            diffs = 0
            for f in fields:
                same = fp_a.get(f) == fp_b.get(f)
                if not same:
                    diffs += 1
                print(f"    {f}: {'SAME' if same else 'DIFFERENT'}")
            print(f"\n  {diffs} of {len(fields)} fields differ between contexts.")

        finally:
            await browser.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/playwright/nodejs/per_context_fingerprint.js
/**
 * PRIVACY RESEARCH USE ONLY
 * Run exclusively in authorized privacy research labs that comply with all applicable laws.
 * See: https://github.com/botswin/BotBrowser/blob/main/DISCLAIMER.md
 */

/**
 * BotBrowser Per-Context Fingerprint Example (Playwright)
 *
 * Creates multiple browser contexts in a single browser instance, each with
 * a different fingerprint profile loaded via BotBrowser.setBrowserContextFlags.
 *
 * Requirements:
 * - BotBrowser binary (not stock Chromium)
 * - At least two .enc profile files
 *
 * Usage:
 *   BOTBROWSER_EXEC_PATH=/path/to/chrome \
 *   PROFILE_A=/path/to/profile-a.enc \
 *   PROFILE_B=/path/to/profile-b.enc \
 *   PROXY_B=socks5://user:pass@host:1080 \
 *   node per_context_fingerprint.js
 */

const { chromium } = require('playwright-core');

(async () => {
  const execPath = process.env.BOTBROWSER_EXEC_PATH;
  const profileA = process.env.PROFILE_A;
  const profileB = process.env.PROFILE_B;
  const proxyB   = process.env.PROXY_B; // optional, e.g. socks5://user:pass@host:1080

  if (!execPath || !profileA || !profileB) {
    console.log('Usage:');
    console.log('  BOTBROWSER_EXEC_PATH=/path/to/chrome \\');
    console.log('  PROFILE_A=/path/to/profile-a.enc \\');
    console.log('  PROFILE_B=/path/to/profile-b.enc \\');
    console.log('  PROXY_B=socks5://user:pass@host:1080 \\');
    console.log('  node per_context_fingerprint.js');
    process.exit(1);
  }

  // Launch browser with profile A as the base profile
  const browser = await chromium.launch({
    executablePath: execPath,
    headless: true,
    ignoreDefaultArgs: [
      '--disable-crash-reporter',
      '--disable-crashpad-for-testing',
      '--disable-gpu-watchdog',
    ],
    args: [
      '--disable-blink-features=AutomationControlled',
      '--disable-audio-output',
      `--bot-profile=${profileA}`,
    ],
  });

  console.log('Browser launched\n');

  // Browser-level CDP session. Required for BotBrowser.* commands.
  // Page-level sessions (page.createCDPSession()) do NOT have access
  // to the BotBrowser domain.
  const browserCDP = await browser.newBrowserCDPSession();

  try {
    // --- Context A: profile A ---
    // 1. Snapshot existing context IDs
    const { browserContextIds: before1 } = await browserCDP.send('Target.getBrowserContexts');
    // 2. Create context via Playwright
    const ctxA = await browser.newContext();
    // 3. Find the new browserContextId
    const { browserContextIds: after1 } = await browserCDP.send('Target.getBrowserContexts');
    const ctxIdA = after1.filter(id => !before1.includes(id))[0];
    // 4. Set flags BEFORE creating any page
    await browserCDP.send('BotBrowser.setBrowserContextFlags', {
      browserContextId: ctxIdA,
      botbrowserFlags: [`--bot-profile=${profileA}`],
    });
    // 5. Now create the page. The renderer starts with the correct flags.
    const pageA = await ctxA.newPage();

    // --- Context B: profile B + optional proxy ---
    const { browserContextIds: before2 } = await browserCDP.send('Target.getBrowserContexts');
    const ctxB = await browser.newContext();
    const { browserContextIds: after2 } = await browserCDP.send('Target.getBrowserContexts');
    const ctxIdB = after2.filter(id => !before2.includes(id))[0];

    const flagsB = [`--bot-profile=${profileB}`];
    if (proxyB) flagsB.push(`--proxy-server=${proxyB}`);

    await browserCDP.send('BotBrowser.setBrowserContextFlags', {
      browserContextId: ctxIdB,
      botbrowserFlags: flagsB,
    });
    const pageB = await ctxB.newPage();

    // Navigate both contexts
    await Promise.all([
      pageA.goto('https://example.com', { waitUntil: 'domcontentloaded' }),
      pageB.goto('https://example.com', { waitUntil: 'domcontentloaded' }),
    ]);

    // Collect fingerprints from each context
    const fpA = await collectFingerprint(pageA);
    const fpB = await collectFingerprint(pageB);

    console.log('Context A fingerprint:');
    printFingerprint(fpA);

    console.log('\nContext B fingerprint:');
    printFingerprint(fpB);

    // Check proxy IP for context B (if proxy was set)
    if (proxyB) {
      console.log('\n--- Proxy check (Context B) ---');
      try {
        await pageB.goto('https://httpbin.org/ip', { timeout: 20000 });
        const ipBody = await pageB.evaluate(() => document.body.innerText);
        const ip = JSON.parse(ipBody).origin;
        console.log(`  Context B exit IP: ${ip}`);
      } catch (e) {
        console.log(`  Proxy check failed: ${e.message}`);
      }
    }

    // Verify isolation
    console.log('\n--- Isolation check ---');
    const fields = ['userAgent', 'platform', 'hardwareConcurrency', 'screenWidth', 'webglRenderer'];
    let differences = 0;
    for (const f of fields) {
      const same = fpA[f] === fpB[f];
      if (!same) differences++;
      console.log(`  ${f}: ${same ? 'SAME' : 'DIFFERENT'}`);
    }
    console.log(`\n${differences} of ${fields.length} fields differ between contexts.`);

  } finally {
    await browser.close();
  }
})();

async function collectFingerprint(page) {
  return page.evaluate(() => {
    const r = {};
    r.userAgent = navigator.userAgent;
    r.platform = navigator.platform;
    r.hardwareConcurrency = navigator.hardwareConcurrency;
    r.deviceMemory = navigator.deviceMemory;
    r.screenWidth = screen.width;
    r.screenHeight = screen.height;
    try {
      const c = document.createElement('canvas');
      const gl = c.getContext('webgl');
      if (gl) {
        const d = gl.getExtension('WEBGL_debug_renderer_info');
        if (d) r.webglRenderer = gl.getParameter(d.UNMASKED_RENDERER_WEBGL);
      }
    } catch (_) {}
    r.timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    return r;
  });
}

function printFingerprint(fp) {
  console.log(`  UA:       ${fp.userAgent}`);
  console.log(`  Platform: ${fp.platform}`);
  console.log(`  HW:       ${fp.hardwareConcurrency} cores, ${fp.deviceMemory}GB`);
  console.log(`  Screen:   ${fp.screenWidth}x${fp.screenHeight}`);
  console.log(`  WebGL:    ${fp.webglRenderer || 'N/A'}`);
  console.log(`  Timezone: ${fp.timezone}`);
}
